![title](images/ebay_logo.png)

### Project introduction

In this project, I'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.

The data dictionary is as follows:

- `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.
- `name` - Name of the car.
- `seller` - Whether the seller is private or a dealer.
- `offerType` - The type of listing
- `price` - The price on the ad to sell the car.
- `abtest` - Whether the listing is included in an A/B test.
- `vehicleType` - The vehicle Type.
- `yearOfRegistration` - The year in which the car was first registered.
- `gearbox` - The transmission type.
- `powerPS` - The power of the car in PS.
- `model` - The car model name.
- `kilometer` - How many kilometers the car has driven.
- `monthOfRegistration` - The month in which the car was first registered.
- `fuelType` - What type of fuel the car uses.
- `brand` - The brand of the car.
- `notRepairedDamage` - If the car has a damage which is not yet repaired.
- `dateCreated` - The date on which the eBay listing was created.
- `nrOfPictures` - The number of pictures in the ad.
- `postalCode` - The postal code for the location of the vehicle.
- `lastSeenOnline` - When the crawler saw this ad last online.

The aim of this project is to clean the data and analyze the included used car listings.

In [1]:
#import the required libraries
import numpy as np
import pandas as pd

In [2]:
#read the file into pandas
autos = pd.read_csv("autos.csv",encoding="Latin-1" )

In [3]:
#summary of the dataframe
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

The dataset shows that there is 5000 rows comprising of integers and objects, which are usually strings. There are also some missing values which we will explore in greater detail a bit later.

In [4]:
#first 5 rows
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

### Cleaning column names

To begin with, we need to change the column headers to snakecase and reword some of the column names based on the data dictionary to be more descriptive.

https://www.geeksforgeeks.org/how-to-rename-columns-in-pandas-dataframe/

In [5]:
#print an array of the existing column names
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
#create a list of snake case column names
col_names = ('date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_PS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_pictures', 'postal_code',
       'last_seen')

In [7]:
#assign the new list to the dataframe
autos.columns = col_names

In [8]:
#print out the new columns
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_PS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

### Initial exploration and cleaning

Now let's do some basic data exploration to determine what other cleaning tasks need to be done. Initially we will look for: - Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis. - Examples of numeric data stored as text which can be cleaned and converted.

The following methods are helpful for exploring the data: - `DataFrame.describe()` (with `include='all'` to get both categorical and numeric columns) - `Series.value_counts()` and `Series.head()` if any columns need a closer look.

In [9]:
#get descriptive statistics for all columns
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-12 16:06:22  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_PS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_pictures   postal_code            last_seen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

There are a couple of things worth noting here;

- `seller` and `offer_type` have mostly 1 value and should probably be dropped
- `price` and `odometer` are strings and need to be converted to integers

Let's start by converting these strings...

https://datatofish.com/string-to-integer-dataframe/

In [10]:
#remove the comma and 'km' text from the values and convert to an integer
autos["odometer"] = autos["odometer"].str.replace(",", "").str.replace("km", "").astype(int)

In [11]:
#remove the $ sign and comma from the values and convert to an integer
autos["price"] = autos["price"].str.replace(",", "").str.replace("$", "").astype(int)

In [12]:
#check the results
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
date_crawled          50000 non-null object
name                  50000 non-null object
seller                50000 non-null object
offer_type            50000 non-null object
price                 50000 non-null int64
abtest                50000 non-null object
vehicle_type          44905 non-null object
registration_year     50000 non-null int64
gearbox               47320 non-null object
power_PS              50000 non-null int64
model                 47242 non-null object
odometer              50000 non-null int64
registration_month    50000 non-null int64
fuel_type             45518 non-null object
brand                 50000 non-null object
unrepaired_damage     40171 non-null object
ad_created            50000 non-null object
num_pictures          50000 non-null int64
postal_code           50000 non-null int64
last_seen             50000 non-null object
dtypes: int64(7), 

In [13]:
#Rename the odometer column
autos.rename(columns={'odometer': 'odometer_km'}, inplace=True)

In [14]:
#check column names
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_PS', 'model',
       'odometer_km', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

### Exploring the odometer and price columns

Let's continue exploring the data, specifically looking for data that doesn't look right. We'll start by analyzing the odometer_km and price columns. Here's the steps we'll take:

- Analyze the columns using minimum and maximum values and look for any values that look unrealistically high or low (outliers) that we might want to remove.

In [15]:
#check to see how many unique values for the price column
autos["price"].unique().shape

(2357,)

In [16]:
#check to see the max value for price
autos["price"].max()

99999999

In [17]:
#check to see the min value for price
autos["price"].min()

0

In [18]:
#check the bottom 5 prices
autos["price"].value_counts().sort_index(ascending=True).head()

0    1421
1     156
2       3
3       1
5       2
Name: price, dtype: int64

In [19]:
#check the top 5 prices
autos["price"].value_counts().sort_index(ascending=False).head()

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
Name: price, dtype: int64

We can see there are some outliers here that we need to deal with. Let's filter out some of these unreasonable prices from the dataset.

For our purposes we'll only keep vehicles that have a price betweeen \\$500 -  \$100,000 dollars

In [20]:
#remove any values <$500 and >$100,000
autos = autos[autos["price"].between(500, 100000)]

In [21]:
#check for max value
autos["price"].max()

99900

In [22]:
#check for min value
autos["price"].min()

500

Great, that all worked fine, now lets do the same for odometer.

In [23]:
#check to see how many unique values for the odometer column
autos["odometer_km"].unique().shape

(13,)

In [24]:
#check to see the max value for odometer
autos["odometer_km"].max()

150000

In [25]:
#check to see the min value for price
autos["odometer_km"].min()

5000

In [26]:
#check the bottom 5 odometer readings
autos["odometer_km"].value_counts().sort_index(ascending=True).head()

5000     607
10000    234
20000    719
30000    759
40000    807
Name: odometer_km, dtype: int64

In [27]:
#check the top 5 odometer readings
autos["odometer_km"].value_counts().sort_index(ascending=False).head()

150000    28695
125000     4836
100000     2029
90000      1676
80000      1384
Name: odometer_km, dtype: int64

The odometer readings vary from 5000 - 150,000kms which is an acceptable range to work from.

### Exploring the date columns

In the dataset there are 5 columns that represent date values;

- `date_crawled`
- `last_seen`
- `ad_created`
- `registration_month`
- `registration_year`

Right now, the `date_crawled`, `last_seen`, and `ad_created` columns are all identified as string values by pandas. Because these three columns are represented as strings, we need to convert the data into a numerical representation so we can understand it quantitatively. The other two columns are represented as numeric values, so we can use methods like `Series.describe()` to understand the distribution without any extra data processing.

Let's first understand how the values in the three string columns are formatted. These columns all represent full timestamp values, like so:

In [28]:
#calculate the distribution of values as percentages for date_crawled, last_seen & ad_created
autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025545
2016-03-06    0.014137
2016-03-07    0.036220
2016-03-08    0.033157
2016-03-09    0.032913
2016-03-10    0.032736
2016-03-11    0.033046
2016-03-12    0.037330
2016-03-13    0.015536
2016-03-14    0.036331
2016-03-15    0.034001
2016-03-16    0.029362
2016-03-17    0.031160
2016-03-18    0.012850
2016-03-19    0.034777
2016-03-20    0.038084
2016-03-21    0.037729
2016-03-22    0.032935
2016-03-23    0.032403
2016-03-24    0.028985
2016-03-25    0.031093
2016-03-26    0.032625
2016-03-27    0.031160
2016-03-28    0.034822
2016-03-29    0.033290
2016-03-30    0.033290
2016-03-31    0.031670
2016-04-01    0.033867
2016-04-02    0.035798
2016-04-03    0.038817
2016-04-04    0.036642
2016-04-05    0.013161
2016-04-06    0.003174
2016-04-07    0.001354
Name: date_crawled, dtype: float64

The distribution looks fairly even with no obvious outliers here, lets now look at `last_seen`.

In [29]:
#calculate the distribution of values as percentages for date_crawled, last_seen & ad_created
autos["last_seen"].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001087
2016-03-06    0.004172
2016-03-07    0.005216
2016-03-08    0.007013
2016-03-09    0.009454
2016-03-10    0.010298
2016-03-11    0.012051
2016-03-12    0.023925
2016-03-13    0.008877
2016-03-14    0.012295
2016-03-15    0.015691
2016-03-16    0.016157
2016-03-17    0.027698
2016-03-18    0.007390
2016-03-19    0.015425
2016-03-20    0.020418
2016-03-21    0.020662
2016-03-22    0.021261
2016-03-23    0.018399
2016-03-24    0.019508
2016-03-25    0.018598
2016-03-26    0.016423
2016-03-27    0.015447
2016-03-28    0.020529
2016-03-29    0.021372
2016-03-30    0.024169
2016-03-31    0.023459
2016-04-01    0.022882
2016-04-02    0.024879
2016-04-03    0.024946
2016-04-04    0.024324
2016-04-05    0.126481
2016-04-06    0.225354
2016-04-07    0.134138
Name: last_seen, dtype: float64

There is alot more activity at the end of 2016 with around 50% occuring in early April 2016. Something worth investigating a bit closer.

In [30]:
#calculate the distribution of values as percentages for date_crawled, last_seen & ad_created
autos["ad_created"].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11    0.000022
2015-08-10    0.000022
2015-09-09    0.000022
2015-11-10    0.000022
2015-12-05    0.000022
2015-12-30    0.000022
2016-01-03    0.000022
2016-01-07    0.000022
2016-01-10    0.000044
2016-01-13    0.000022
2016-01-14    0.000022
2016-01-16    0.000022
2016-01-22    0.000022
2016-01-27    0.000067
2016-01-29    0.000022
2016-02-01    0.000022
2016-02-02    0.000044
2016-02-05    0.000044
2016-02-07    0.000022
2016-02-08    0.000022
2016-02-09    0.000022
2016-02-11    0.000022
2016-02-12    0.000044
2016-02-14    0.000044
2016-02-16    0.000022
2016-02-17    0.000022
2016-02-18    0.000044
2016-02-19    0.000067
2016-02-20    0.000044
2016-02-21    0.000044
                ...   
2016-03-09    0.033024
2016-03-10    0.032469
2016-03-11    0.033357
2016-03-12    0.037108
2016-03-13    0.016978
2016-03-14    0.034911
2016-03-15    0.033779
2016-03-16    0.029850
2016-03-17    0.030805
2016-03-18    0.013494
2016-03-19    0.033646
2016-03-20    0.038217
2016-03-21 

Notice how ad creations really picked up in March 2016. Another investigation point.

In [31]:
#check the distribution of registration_year
autos["registration_year"].describe()

count    45058.000000
mean      2005.063918
std         89.689852
min       1000.000000
25%       2000.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

### Dealing with incorrect registration year data

One thing that stands out from the exploration we did in the last screen is that the `registration_year` column contains some odd values:

- The minimum value is `1000`, before cars were invented
- The maximum value is `9999`, many years into the future

Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

Let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely, or if we need more custom logic.

In [32]:
#calculate the % of incorrect values
sum((autos["registration_year"] < 1900) | (autos["registration_year"] > 2016)) / len(autos)

0.03937147676328288

Given that the amount of values only make up ~4% of the `registration_year` column, let's remove these from the dataframe. 

In [33]:
#remove incorrect values and reassign to the dataframe
autos = autos[autos["registration_year"].between(1900, 2016)]

In [34]:
#test the results
autos["registration_year"].describe()

count    43284.000000
mean      2003.225349
std          7.065511
min       1910.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

### Exploring price by brand

In [35]:
#explore unique values in the brand column
autos["brand"].value_counts().head(10)

volkswagen       9180
bmw              4980
mercedes_benz    4427
opel             4347
audi             3920
ford             2822
renault          1902
peugeot          1272
fiat             1019
seat              775
Name: brand, dtype: int64

Let's look at the top 10 most popular brands and see what the average price is for each brand.

In [36]:
#Create a list of unique brands
brands = autos["brand"].value_counts().head(10).index

In [37]:
#create an empty dictionary
brands_mean_value = {}

#use a for loop to iterate through selected brands
for brand in brands:
    #filter through brands
    selected_brands = autos[autos["brand"] == brand]
    #apply the mean function
    mean = selected_brands["price"].mean()
    #append results back tot he dictionary
    brands_mean_value[brand] = mean

In [38]:
#check the results
brands_mean_value

{'audi': 9571.457397959184,
 'bmw': 8447.069879518072,
 'fiat': 3256.152109911678,
 'ford': 4247.120481927711,
 'mercedes_benz': 8666.677208041563,
 'opel': 3394.0395675178283,
 'peugeot': 3360.9205974842766,
 'renault': 2819.059411146162,
 'seat': 4810.883870967742,
 'volkswagen': 5783.622984749455}

Not suprsing in the top 6 brands, there's a distinct price gap.

- Audi, BMW and Mercedes Benz are more expensive
- Ford and Opel are less expensive 
- Volkswagen is in between

For the top 6 brands, let's use aggregation to understand the average mileage for those cars and if there's any visible link with mean price.

### Storing aggregate data in a dataframe

In [39]:
#Create a list of unique brands
brands = autos["brand"].value_counts().head(10).index

In [40]:
#create an empty dictionary
od_mean_value = {}

#use a for loop to iterate through selected brands
for brand in brands:
    #filter through brands
    selected_brands = autos[autos["brand"] == brand]
    #apply the mean function
    mean = selected_brands["odometer_km"].mean()
    #append results back tot he dictionary
    od_mean_value[brand] = mean

In [41]:
od_mean_value

{'audi': 128941.32653061225,
 'bmw': 132928.71485943775,
 'fiat': 114416.09421000982,
 'ford': 123520.55279943303,
 'mercedes_benz': 131083.12627061215,
 'opel': 128012.42236024844,
 'peugeot': 126073.11320754717,
 'renault': 126351.20925341746,
 'seat': 120058.06451612903,
 'volkswagen': 128234.74945533769}

In [42]:
#convert price dictionary to a series
price_series = pd.Series(brands_mean_value)

In [43]:
#convert odometer dictionary to a series
od_series = pd.Series(od_mean_value)

In [45]:
#create a new dataframe from the price series
mean_df = pd.DataFrame(price_series, columns=["mean_price"])

In [46]:
#check results
mean_df

mean_price
audi           9571.457398
bmw            8447.069880
fiat           3256.152110
ford           4247.120482
mercedes_benz  8666.677208
opel           3394.039568
peugeot        3360.920597
renault        2819.059411
seat           4810.883871
volkswagen     5783.622985

In [47]:
#assign the odometer series to the new dataframe
mean_df["mean_odometer"] = od_series

In [48]:
#check the results
mean_df

mean_price  mean_odometer
audi           9571.457398  128941.326531
bmw            8447.069880  132928.714859
fiat           3256.152110  114416.094210
ford           4247.120482  123520.552799
mercedes_benz  8666.677208  131083.126271
opel           3394.039568  128012.422360
peugeot        3360.920597  126073.113208
renault        2819.059411  126351.209253
seat           4810.883871  120058.064516
volkswagen     5783.622985  128234.749455

Now we have merged the two columns into a dataframe, we can draw a couple of conclusions;

- The brands with the highest mean price also have the highest mean odometer reading (Mercedes & BMW)
- The lower priced cars have the lower odometer readings
- There appears to be no negative correlation between a high odometer reading and a lower car price
- Expensive brands tend to hold their value better than cheaper cars

### Next steps...

In this project, we practiced applying a variety of pandas methods to explore and understand a data set on car listings. Here are some next steps to consider to further our analsis:

Data cleaning next steps:
- Identify categorical data that uses german words, translate them and map the values to their english counterparts
- Convert the dates to be uniform numeric data, so `"2016-03-21"` becomes the integer `20160321`.
- See if there are particular keywords in the name column that you can extract as new columns

Analysis next steps:
- Find the most common brand/model combinations
- Split the `odometer_km` into groups, and use aggregation to see if average prices follows any patterns based on the milage.
- How much cheaper are cars with damage than their non-damaged counterparts?